### PR TITLE
Building casual attention mask

### PR DESCRIPTION
**Pull request by iejMac**

### CODE REVIEW

File: src/open_clip/hf_model.py [hunk 1]
```diff
@@ -48,10 +49,22 @@ class PreTrainedTextEncoder(nn.Module):
...
             self, 
             model_name_or_path:str,
-            embedding_dim:int,
+
+            context_length:int = 77,
+            vocab_size:int = 49408,
+            width:int = 512,
+            heads:int = 8,
+            layers:int = 12,
+            output_dim:int = 512,
```
Comment (arampacha) [inline src/open_clip/hf_model.py]: vocab_size, heads, layers is given for pretrained model, so we really only need  an argument for the expected output embedding dim. We can also add an optional `config` input default to none, to enable instantiating untrained HF transformer models. For this scenario different models will require different args, so it's better to pass `config`.

File: src/open_clip/hf_model.py [hunk 2]
```diff
@@ -61,17 +74,31 @@ def __init__(
...
+        return mask
+
     def forward(self, x:TensorType) -> TensorType:
         # TODO: add attention mask?
-        out = self.transformer(input_ids=x)
+        pad = torch.full((x.shape[0], self.context_length), self.config.pad_token_id)
+        pad[:, :x.shape[1]] = x
+        x = pad
+
+        out = self.transformer(input_ids=x, attention_mask=self.attn_mask[:x.shape[0]])
```
Comment (arampacha) [inline src/open_clip/hf_model.py]: Don't we have x padded already in shape [bs, context_len] at this stage? Or is it a ragged tensor?
If x is padded already, one can make attention mask as `(x != self.config.pad_token_id).long()`
Otherwise it can be better to instantiate `self.tokenizer = AutoTokenizer.from_pretrained(...)` and use `self.tokenizer.pad`

File: src/open_clip/hf_model.py (hunk 1 above)
Comment (iejMac) [reply src/open_clip/hf_model.py]: right so we want to add a config of type PretrainedConfig and if that argument is passed and hence isn't None we want to instantiate the model like:
```
self.transformer = AutoModel.from_pretrained(config=self.config)
```
yes?

Comment (arampacha) [reply src/open_clip/hf_model.py]: yes, this should work

Comment (arampacha) [reply src/open_clip/hf_model.py]: actually `from_pretrained` will throw error, there is `_from_config` method that should fit in this case

Comment (iejMac) [reply src/open_clip/hf_model.py]: Yup, that's what I used. You can check out the latest changes and LMK if it looks good to merge for you
